Fix how city decorator adds a default for detector_db

jmbenlloch opened the pull request:
In #618 a new parameter was introduced to choose which database to read depending on the detector. Then, all IC cities needed a `detector_db` parameter to read the correct data. To keep backward compatibility a default value for `detector_db` was needed even if that variable was not in the config file to be run.

To do so, a change was introduced in the decorator `@city`. The problem now, as @jmunozv has reported, is that there are some functions in clients of IC (FANAL) that use `@city` but do not read the database at all, so there is not point in giving them `detector_db`.

The change introduced in this PR uses python reflection mechanisms to pass a default `detector_db` value only to those cities that need it.

Maybe @paolafer or @jmunozv can review this.

jmbenlloch committed:
Fix how city decorator adds a default for detector_db

All IC cities needs a detector_db parameter to read the correct data
depending on the detector each job is using. But there are some
functions outside IC (in FANAL) that are also cities and does not
expect a detector_db parameter.

This parameter will have only a default value only if the decorated
function expect it.

jacg commented on invisible_cities/cities/components.py:
```diff
@@ -72,7 +73,8 @@ def proxy(**kwds):
         # For backward-compatibility we set NEW as the default DB in
         # case it is not defined in the config file
-        if 'detector_db' not in kwds:
+#        if 'detector_db' not in kwds:
```
Commented line needs removing.

jmbenlloch replied:
Done, I have updated the commit.

jmbenlloch committed:
Add tests to check expected behaviour with detector_db

There are thre situations that are now tested:

- Function expects detector_db and that value is not defined in the
 config file: A default value is passed by @city (NEW detector).

- Function expects detector_db and that value is defined in the config
  file: The value is not overwritten by @city and is properly passed.

- Function does not expect detector_db: @city does not pass it.

paolafer reviewed (approved):
This PR makes possible to use the new version of IC, with more than one detector database, with fanal and other clients which know nothing about databases. A few tests have been added to show the behaviour of the code in different cases, therefore I approve this PR.
